state/application.go

```diff
@@ -3124,6 +3132,83 @@ func allUnits(st *State, application string) (units []*Unit, err error) {
 	return units, nil
 }
 
+// GetUnitAttachmentInfos returns a list of UnitAttachmentInfo for this application.
+func (a *Application) GetUnitAttachmentInfos(statuses []status.Status) (unitAttachmentInfos []UnitAttachmentInfo, err error) {
+	logger.Warningf("jneo8 unitCount: %d scale: %d", a.doc.UnitCount, a.doc.DesiredScale)
+	// In case of scale down, there won't have attachment info.
+	if a.doc.DesiredScale < a.doc.UnitCount {
+		return unitAttachmentInfos, nil
+	}
+
+	unitIds := []string{}
+
+	// When deploying an application, UnitCount will match DesiredScale (which defaults to 1).
+	// Note: Attaching storage is not allowed during deployment if DesiredScale > 1, so that case is ignored.
+	// When using add-unit, DesiredScale will be greater than UnitCount during the initial part of the lifecycle.
+	if (a.doc.UnitCount == 1 && a.doc.DesiredScale == 1) || (a.doc.UnitCount < a.doc.DesiredScale) {
+		for i := a.doc.UnitCount; i <= a.doc.DesiredScale; i++ {
+			unitIds = append(unitIds, strconv.Itoa(i))
+		}
+	}
+
+	if len(unitIds) == 0 {
+		return unitAttachmentInfos, nil
+	}
+
+	idReg := "(" + strings.Join(unitIds, "|") + ")"
```

Comment:
We should include in the comment at the top of this method that the method is called in the same worker loop that updates the scale so that we can avoid doing consistency checks since provisioning and scale updates are interleaved.